File: 2.DataStructure/Python/Heaps/heapsort.py
# Modified percolateDown
def percolateDown(A, first, last):
    largest = 2 * first + 1
    while largest <= last:
        # right child exist and is larger than left minChild
        if(largest < last) and (A[largest] < A[largest + 1]):
            largest += 1
        # right child is larger than parent
        if A[largest] > A[first]:
            swap(A, largest, first)
            # move down to largest minChild
            first = largest
            largest = 2 * first + 1
        else:
            return


def swap(A, x, y):
    # temp = A[x]
    # A[x] = A[y]
    # A[y] = temp
    A[x]=A[x]^A[y]
    A[y]=A[y]^A[x]
    A[x]=A[x]^A[y]


def heapSort(A):
    length = len(A) - 1
    leastParent = length // 2
    for i in range(leastParent, -1, -1):
        percolateDown(A, i, length)
    
    print(*A)

    for i in range(length, 0, -1):
        if A[0] > A[i]:
            swap(A, 0, i)
            percolateDown(A, 0, i - 1)

print("Inside HeapSort")
a = [10, 78, 34, 56, 32, 99]
heapSort(a)
for i in a:
    print(i)
print("Exiting HeapSort")

File: 1.Algorithms/C++/Sorting/heap_sort.cpp
#include <bits/stdc++.h>
using namespace std;

void heapify(vector<int> &A, int n, int index)
{
    int largest = index;       // Index
    int left = 2 * index + 1;  // Left Node
    int right = 2 * index + 2; // Right Node

    if (left < n && A[left] > A[largest])
        largest = left;

    if (right < n && A[right] > A[largest])
        largest = right;

    if (largest != index)
    {
        swap(A[largest], A[index]);
        heapify(A, n, largest);
    }
}

void heap_Sort(vector<int> &A)
{
    int n = int(A.size());

    for (int i = (n >> 1) - 1; i >= 0; i--)
    {
        heapify(A, n, i);
    }

    for (int i = n - 1; i >= 0; i--)
    {
        swap(A[0], A[i]);
        heapify(A, i, 0);
    }
}

int main()
{
    vector<int> A = {1, 7, 2, 5, 2, 3, 9};
    heap_Sort(A);

    for (auto i : A)
        cout << i << " ";
}